### PR TITLE
Add in negative size checks for columns

### LINKED ISSUE #12116: [BUG] Allocating columns with small negative lengths succeeds
**Describe the bug**
This is directly related to https://github.com/rapidsai/rmm/issues/1156. I recently tried to add in some extra error checking on top of `cudf::strings::concatenate` because I wanted to be ready in case https://github.com/rapidsai/cudf/issues/12087 didn't work out. As a part of this I tested overflowing by just a very small amount 4 characters more than the maximum amount.  To my shock the call to concatenate passed but trying to convert that result to a `column_view` threw an exception saying that a `column_view` cannot have a negative length. It appears that the negative allocation is turned into a very large positive allocation because C++ inserts in an implicit cast with no warnings, then https://github.com/rapidsai/rmm/issues/1156 kicks in and causes the allocation to be converted to 0 which returns a `nullptr` and succeeds. My concern is not so much with strings directly.  Even if https://github.com/rapidsai/cudf/issues/12087 is fixed the same kind of thing can happen for any allocation that overflows by a small amount.
 
**Steps/Code to reproduce bug**
Allocate a `numeric_column` of `INT8` with a length of `-4`. It will work and there will be no exceptions until you try to turn it into a column_view, which might not happen if you are trying to use the column directly to pull out raw pointers/etc.

**Expected behavior**
I would like very much if it were not possible to allocate negative length columns. If we have the assertion on a `column_view` why do we not have the same thing for a `column`? This way even if the RMM issue is fixed don't have to wonder about why we are trying to allocate really large allocations from RMM that failed, but instead we can take the next step and know that something overflowed and we need to look at who is trying to allocate a negative `column`.

### PR DESCRIPTION
**Pull request by revans2**
## Description
This fixes #12116
This just adds in a few checks for negative sizes to avoid any issues with rounding errors and also helps us detect errors sooner. It will not fix small negative allocations for device buffers directly.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/column/column.hpp [hunk 1]
```diff
@@ -133,6 +133,7 @@ class column {
       _null_count{null_count},
       _children{std::move(children)}
   {
+    CUDF_EXPECTS(size >= 0, "Column size cannot be negative.");
```
[the remark this answers is not in the document]
Comment (revans2) [reply cpp/include/cudf/column/column.hpp:136]: That is correct, but I added the other checks as a way to skip allocating negative data and null mask buffers with a negative size if we know early on that it is not needed. If you want me to remove the other checks I will.

[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/include/cudf/column/column.hpp:136]: Nope, that makes sense. We should leave those in.
The other thing to do here is to add `@throw` tag to the doxygen comments for each of these
```
* @throws cudf::logic_error if `size < 0`
```
The `column_factories` ones go into `cpp/include/cudf/column/column_factories.hpp`

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

**Comment by revans2** [comment]
@davidwendt thanks for the review. I added in the doxegen tags.

**Comment by davidwendt** [comment]
> @davidwendt thanks for the review. I added in the doxegen tags.

Can you move the `@throw` tags up in the doxygen next to the other `@throw` tags?
Guidelines for reference: https://github.com/rapidsai/cudf/blob/branch-22.12/cpp/doxygen/developer_guide/DOCUMENTATION.md#function-parameters

**Comment by revans2** [comment]
@davidwendt sorry about getting the placement wrong. Hopefully it is fixed now.

**Review by davidwendt — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Comment by revans2** [comment]
@gpucibot merge

**Pull request merged**